File: summary/ruby/19_knapsack.rb
def knapsack(weights, values, capacity)
  n = weights.length
  dp = Array.new(n + 1) { Array.new(capacity + 1, 0) }

  (1..n).each do |i|
    (1..capacity).each do |w|
      if weights[i - 1] <= w
        dp[i][w] = [dp[i - 1][w], values[i - 1] + dp[i - 1][w - weights[i - 1]]].max
      else
        dp[i][w] = dp[i - 1][w]
      end
    end
  end

  dp[n][capacity]
end

# Example usage:
weights = gets.chomp.split.map(&:to_i)
values = gets.chomp.split.map(&:to_i)
capacity = gets.to_i
max_value = knapsack(weights, values, capacity)
puts "Maximum value that can be accommodated: #{max_value}"

File: summary/python/19_knapsack.py
def knapsack(weights, values, capacity):
    n = len(weights)
    dp = [[0 for _ in range(capacity + 1)] for _ in range(n + 1)]

    for i in range(1, n + 1):
        for w in range(1, capacity + 1):
            if weights[i - 1] <= w:
                dp[i][w] = max(dp[i - 1][w], values[i - 1] + dp[i - 1][w - weights[i - 1]])
            else:
                dp[i][w] = dp[i - 1][w]

    return dp[n][capacity]

# Example usage:
weights = list(map(int, input().split()))
values = list(map(int, input().split()))
capacity = int(input())
max_value = knapsack(weights, values, capacity)
print(f"Maximum value that can be accommodated: {max_value}")
